gst/nnstreamer/tensor_sparse/tensor_sparse_util.c

```diff
@@ -0,0 +1,219 @@
+
+/**
+ * @brief Make dense tensor with input sparse tensor.
+ * @param[in,out] meta tensor meta structure to be updated
+ * @param[in] in pointer of input sparse tensor data
+ * @return  pointer of Gstmemory with dense tensor data or NULL on error. Caller should handle this newly allocated memory.
+ */
+extern GstMemory *
+gst_tensor_sparse_to_dense (GstTensorMetaInfo * meta, gpointer in)
+{
+  guint i, nnz;
+  guint8 *output, *input;
+  guint *indices;
+  gsize output_size, element_size = gst_tensor_get_element_size (meta->type);
+
+  meta->format = _NNS_TENSOR_FORMAT_STATIC;
+
+  output_size = gst_tensor_meta_info_get_data_size (meta);
+  output = (guint8 *) g_malloc0 (output_size);
+
+  nnz = meta->sparse_info.nnz;
+  input = (guint8 *) in + gst_tensor_meta_info_get_header_size (meta);
+  indices = ((guint *) ((guint8 *) input + element_size * nnz));
+
+  for (i = 0; i < nnz; ++i) {
+    switch (meta->type) {
+      case _NNS_INT32:
+        ((int32_t *) output)[indices[i]] = ((int32_t *) input)[i];
+        break;
+      case _NNS_UINT32:
+        ((uint32_t *) output)[indices[i]] = ((uint32_t *) input)[i];
+        break;
+      case _NNS_INT16:
+        ((int16_t *) output)[indices[i]] = ((int16_t *) input)[i];
+        break;
+      case _NNS_UINT16:
+        ((uint16_t *) output)[indices[i]] = ((uint16_t *) input)[i];
+        break;
+      case _NNS_INT8:
+        ((int8_t *) output)[indices[i]] = ((int8_t *) input)[i];
+        break;
+      case _NNS_UINT8:
+        ((uint8_t *) output)[indices[i]] = ((uint8_t *) input)[i];
+        break;
+      case _NNS_FLOAT64:
+        ((double *) output)[indices[i]] = ((double *) input)[i];
+        break;
+      case _NNS_FLOAT32:
+        ((float *) output)[indices[i]] = ((float *) input)[i];
+        break;
+      case _NNS_INT64:
+        ((int64_t *) output)[indices[i]] = ((int64_t *) input)[i];
+        break;
+      case _NNS_UINT64:
+        ((uint64_t *) output)[indices[i]] = ((uint64_t *) input)[i];
+        break;
+      default:
+        nns_loge ("Error occured during get tensor value");
+        g_free (output);
+
+        return NULL;
+    }
+  }
+
+  return gst_memory_new_wrapped (0, output, output_size, 0,
+      output_size, output, g_free);
+}
+
+/**
+ * @brief Make sparse tensor with input dense tensor.
+ * @param[in,out] meta tensor meta structure to be updated
+ * @param[in] in pointer of input dense tensor data
+ * @return pointer of GstMemory with sparse tensor data or NULL on error. Caller should handle this newly allocated memory.
+ */
+extern GstMemory *
+gst_tensor_sparse_from_dense (GstTensorMetaInfo * meta, gpointer in)
+{
+  guint i, nnz = 0;
+  guint8 *output;
+  tensor_type data_type;
+  void *values;
+  guint *indices;
+  gsize output_size;
+  gsize header_size = gst_tensor_meta_info_get_header_size (meta);
+  gsize element_size = gst_tensor_get_element_size (meta->type);
+  gulong element_count = gst_tensor_get_element_count (meta->dimension);
```

Comment:
Better to check those variables' validity before starting?

Reply:
Right, I added lines to check `element_size` and `element_count`. Thanks!